File: app/controllers/auth_controller.py
from flask import jsonify
from flask_jwt_extended import create_access_token
from sqlalchemy import func
from sqlalchemy.exc import IntegrityError

from app.extensions import db
from app.models.user_model import User


def _validate_auth_payload(data, is_register=False):
    errors = []
    if is_register:
        if not data.get("email"):
            errors.append("email is required.")
        if not data.get("password"):
            errors.append("password is required.")
        if not data.get("full_name"):
            errors.append("full_name is required.")
        if data.get("role") not in {"user", "employer"}:
            errors.append("role is required and must be either user or employer.")
    else:
        if not data.get("email"):
            errors.append("email is required.")
        if not data.get("password"):
            errors.append("password is required.")
    return errors


def register(data):
    errors = _validate_auth_payload(data, is_register=True)
    if errors:
        return jsonify({"errors": errors}), 400

    email = str(data["email"]).strip().lower()
    full_name = str(data["full_name"]).strip()
    if not email or not full_name:
        return jsonify({"errors": ["email and full_name cannot be blank."]}), 400

    if User.query.filter(func.lower(User.email) == email).first():
        return jsonify({"error": "Email already registered."}), 409

    user = User(
        email=email,
        full_name=full_name,
        role=data["role"],
    )
    user.set_password(data["password"])
    db.session.add(user)
    try:
        db.session.commit()
    except IntegrityError:
        db.session.rollback()
        return jsonify({"error": "Email already registered."}), 409
    except Exception:
        db.session.rollback()
        return jsonify({"error": "Registration failed. Please try again."}), 500

    token = create_access_token(identity=str(user.id))
    return jsonify({
        "message": "Registered successfully.",
        "access_token": token,
        "user": user.to_dict(viewer_id=user.id, viewer_role=user.role, include_skills=True),
    }), 201


def login(data):
    errors = _validate_auth_payload(data)
    if errors:
        return jsonify({"errors": errors}), 400

    email = str(data["email"]).strip().lower()
    user = User.query.filter(func.lower(User.email) == email).first()
    if not user or not user.check_password(data["password"]):
        return jsonify({"error": "Invalid email or password."}), 401

    if not user.is_active:
        return jsonify({"error": "Account suspended."}), 403

    token = create_access_token(identity=str(user.id))
    return jsonify({
        "access_token": token,
        "user": user.to_dict(viewer_id=user.id, viewer_role=user.role, include_skills=True),
    }), 200


def get_me(user_id):
    user = User.query.get(int(user_id))
    if not user:
        return jsonify({"error": "User not found."}), 404
    if not user.is_active:
        return jsonify({"error": "Account suspended."}), 403
    return jsonify({
        "user": user.to_dict(viewer_id=user.id, viewer_role=user.role, include_stats=True, include_skills=True),
    }), 200

File: src/controllers/authController.js
'use strict';
const mongoose = require('mongoose');
const jwt = require('jsonwebtoken');
const passwordUtil = require('../utils/password');
const { User, UserSkill, Skill, WorkSample, Contract, Review, publicDoc } = require('../models');
const { jwtSecret, jwtExpiresMinutes } = require('../config');

const tokenFor = id => jwt.sign({}, jwtSecret, { subject: String(id), expiresIn: `${jwtExpiresMinutes}m` });
async function serializeUser(user, { self = false, stats = false, skills = false } = {}) {
  const data = publicDoc(user);
  if (!self) for (const key of ['phone_number','phone_verified_at','email_verified_at','address_line1','address_line2','address_city','address_region','address_postal_code','identity_rejection_reason']) delete data[key];
  if (self) {
    data.email_verified_for_identity = Boolean(user.email_verified_at);
    data.phone_verified_for_identity = Boolean(user.phone_verified_at);
  }
  if (skills) {
    const links = await UserSkill.find({ user_id: user.id }).lean();
    data.user_skills = await Promise.all(links.map(async link => ({ ...publicDoc(link), skill: publicDoc(await Skill.findOne({ id: link.skill_id })), ai_reviewed: Boolean(await WorkSample.exists({ user_skill_id: link.id, verification_status: 'plausible' })) })));
  }
  if (stats) {
    data.completed_project_count = await Contract.countDocuments({ assigned_member_id: user.id, status: 'completed' });
    const rating = await Review.aggregate([{ $match: { member_id: user.id } }, { $group: { _id: null, avg: { $avg: '$rating' } } }]);
    data.rating = rating[0] ? Math.round(rating[0].avg * 100) / 100 : 0;
  }
  return data;
}
async function register(req, res, next) {
  try {
    const data = req.body || {}, errors = [];
    const full_name = data.full_name ?? data.fullName;
    const role = data.role ?? data.accountType;
    if (!data.email) errors.push('email is required.'); if (!data.password) errors.push('password is required.');
    if (!full_name) errors.push('full_name is required.'); if (!['user','employer'].includes(role)) errors.push('role is required and must be either user or employer.');
    if (errors.length) return res.status(400).json({ errors });
    const email = String(data.email).trim().toLowerCase(), normalizedFullName = String(full_name).trim();
    if (!email || !normalizedFullName) return res.status(400).json({ errors: ['email and full_name cannot be blank.'] });
    if (mongoose.connection.readyState !== 1) return res.status(503).json({ error: 'Database is not connected. Please try again shortly.' });
    if (!jwtSecret) return res.status(500).json({ error: 'Server authentication configuration error.' });
    if (await User.exists({ email })) return res.status(409).json({ error: 'Email already registered.' });
    const password = await passwordUtil.hash(data.password);
    const user = await User.create({ email, full_name: normalizedFullName, role, password });
    return res.status(201).json({ message: 'Registered successfully.', access_token: tokenFor(user.id), user: await serializeUser(user, { self: true, skills: true }) });
  } catch (err) {
    console.error(`Registration failed: ${err.message}`);
    console.error(err.stack);
    if (err.code === 11000) {
      const duplicateEmail = err.keyPattern?.email || err.keyValue?.email;
      return res.status(409).json({ error: duplicateEmail ? 'Email already registered.' : 'Registration conflicts with an existing account.' });
    }
    if (err.name === 'ValidationError') {
      const errors = Object.values(err.errors).map(error => error.message);
      return res.status(400).json({ errors });
    }
    if (err.name === 'MongoServerSelectionError' || err.name === 'MongoNetworkError' || /buffering timed out|database connection/i.test(err.message)) {
      return res.status(503).json({ error: 'Database connection error. Please try again shortly.' });
    }
    return next(err);
  }
}
async function login(req, res, next) {
  try {
    const data = req.body || {}, errors = [];
    if (!data.email) errors.push('email is required.'); if (!data.password) errors.push('password is required.');
    if (errors.length) return res.status(400).json({ errors });
    const user = await User.findOne({ email: String(data.email).trim().toLowerCase() });
    if (!user || !(await passwordUtil.verify(data.password, user.password))) return res.status(401).json({ error: 'Invalid email or password.' });
    if (!user.is_active) return res.status(403).json({ error: 'Account suspended.' });
    return res.json({ access_token: tokenFor(user.id), user: await serializeUser(user, { self: true, skills: true }) });
  } catch (e) { next(e); }
}
async function me(req, res, next) { try { if (!req.user.is_active) return res.status(403).json({ error: 'Account suspended.' }); res.json({ user: await serializeUser(req.user, { self: true, stats: true, skills: true }) }); } catch (e) { next(e); } }
module.exports = { register, login, me, serializeUser };
